### PR TITLE
[gazebo_ros] Changed the spawn model methods to spawn also lights.

### PR DESCRIPTION
**Pull request by aambrosano**
Created services for deleting lights, and getting and settings lights' properties.
**The code proposed in this commit is being ported from the codebase developed within the Human Brain Project, Subproject 10 (Neurorobotics), as part of the Neurorobotics Platform, by the Neurorobotics Platform team.**

### CODE REVIEW

**Comment by j-rivero** [comment]
Thanks for the pull request Alessandro. I see two different things in it:

1. Support for spawning lights
1. Rename model* fields and functions to entity*

I see no problem for merging the changes on 1 but merging the changes in 2 will break code wrote for Kinetic, which I think it is not a good idea. We were discussion about changes [in this thread]() recently. If you don't mind, I would recommend: create one pull request to implement the light feature. Another pull request to implement the change from model to entity but do not remove the \*model\* functions and messages just add the \*entity\* ones, move the functionality to there and make the old \*model\* functions to call the new \*entity\* functions together with a deprecation warning, this way we can consider them to go into Kinetic.

**Comment by aambrosano** [comment]
Hi Jose Luis, thanks for your comments. This commit should do for the first part.

File: gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h
Comment (j-rivero) [inline gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h]: can we respect here the signature of the function that was previously in place?

File: gazebo_msgs/srv/SpawnModel.srv
Comment (j-rivero) [inline gazebo_msgs/srv/SpawnModel.srv]: message structure and filed names are important not to break the backwards compatibility, can we keep the same name in the message than before?

**Comment by aambrosano** [comment]
Hi Jose Luis, is there any problem with this pull request? Do you want me to remove this pull request and create a new one from scratch, or is just removing the extra code renaming functions (like I did) enough for the first step you suggested? Thank you.

**Comment by j-rivero** [comment]
> Hi Jose Luis, is there any problem with this pull request? Do you want me to remove this pull request and create a new one from scratch, or is just removing the extra code renaming functions (like I did) enough for the first step you suggested? Thank you.

Oh, I wrote the comments about the code but forgot to press the proper "submit review" button, sorry. They should be now in place.

File: gazebo_msgs/srv/SpawnModel.srv
Comment (davetcoleman) [inline gazebo_msgs/srv/SpawnModel.srv]: Seems unnecessary to break a ROS message type mid-release - if I'm not mistaken this would change the md5 sum and is not API compatible. Can you just change the comment?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Unnecessary: default constructor "Constructs an empty string, with a length of zero characters."

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: no need to start using ``this->`` here

File: gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h
Comment (davetcoleman) [inline gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h:157]: do you mind adding spaces between parameters?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp:186]: how about ``light_create_pub_`` so that it matches better ``light_modify_pub_``?

Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp:186]: and put light at the start of the topic name too, so it matches

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: this would break API right??

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: model still seems a better name for a URDF to me - it is not referring to lights here

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp:616]: same comment

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: ROS style guidelines have line break for brackets

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: is ``std::string`` necessary for all the concatenations?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp:2149]: model still seems like better comment here

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp:2408]: is it possible there is both a model and a light with the same name? this conditional would not catch that error

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [inline gazebo_ros/src/gazebo_ros_api_plugin.cpp]: i think we shouldn't support versions older than 6 for the kinetic branch, see https://github.com/ros-simulation/gazebo_ros_pkgs/pull/542

**Review by davetcoleman — CHANGES_REQUESTED** [review]
Great contribution, I almost always have to add a light manually when I launch Gazebo so I can see whats going on in the dark environment!

**Comment by davetcoleman** [comment]
i just noticed @j-rivero comments, and i agree with them

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: I agree, but I just put back the code that was already there, if you check the overall diff you shouldn't see any reference to robot_namespace_ at all.

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Same here

File: gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h
Comment (aambrosano) [reply gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h:157]: I don't, actually I would rather put them, but the line just above mine doesn't have any space between parameters.
I was wondering what the standard is when writing new functions, and it seems like throughout the code this happens many times...

File: gazebo_msgs/srv/SpawnModel.srv
Comment (aambrosano) [reply gazebo_msgs/srv/SpawnModel.srv]: The message has been put back to its original status, please check the overall changes of this pull request

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Oops, this has been left back, my bad

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp:186]: This is actually made for matching the Gazebo communication channel's name, as it's happening for all the other publishers. But if you think light_create_pub_ is a better idea I will change it.

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: In both cases (spawnURDFModel and spawnSDFModel), you can pass a XML definition of a light instead of a model. That's why the comments were changed from "model" to a more generic "entity", and I initially changed the functions names as well initially (to spawnURDFEntity and spawnSDFEntity).

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp:616]: Same here

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: I thought it was, but I was wrong.

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Ok, though there's some "#if GAZEBO_MAJOR_VERSION >= 3" around, I think they should be removed as well. Should I do this in this same pull request?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: I'm going to fix this in the next patchset

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp:2408]: I don't need to catch that particular case, what is being done here is:
- if the entity to spawn is a light, make sure there's no light with the same name
- if it's a model, then check there's no model with the same name
All the other cases should be fine. Lights and models with the same name are allowed AFAIK, as they're saved in different data structures.

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Ah yes, I did a code review on a single commit after clicking an email link, my bad.

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp:186]: Ok, your call

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: then can you reword the whole status message so that it does not just mention "URDF"?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: Different pull request, thanks for helping with code maintenance!

File: gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h
Comment (davetcoleman) [reply gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h:157]: the [roscpp style guidelines](http://wiki.ros.org/CppStyleGuide#Function_.2BAC8_Methods) don't specify this, so I guess my comment is ungrounded. but it does bother me that we are inconsistent throughout this repo on the style. as a moveit maintainer, i have setup clangformat to ensure style is automated and consistent, perhaps we should add that here (its not that hard) @j-rivero ?

File: gazebo_ros/src/gazebo_ros_api_plugin.cpp
Comment (aambrosano) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: How? If I understand correctly the old way to define lights and the SDF one are quite different. I mean, old Gazebo versions could still have a light in a world definition, are you saying that that <light> tag is not URDF?

Comment (davetcoleman) [reply gazebo_ros/src/gazebo_ros_api_plugin.cpp]: A URDF cannot specify lights, so there is not need to call the URDF an "entity" in this message unless I'm missing something.

**Comment by davetcoleman** [comment]
@aambrosano I'm excited to use this feature, do you have example roslaunch code for adding a light to the scene?

**Comment by aambrosano** [comment]
@davetcoleman I didn't prepare any roslaunch code but you can try like this:
- Run `rosrun gazebo_ros gazebo`
- Copy this code on your shell
<pre>
rosservice call /gazebo/spawn_sdf_model "model_name: 'dirlight1'
model_xml: '&lt;sdf version="1.5"&gt;
    &lt;light type="directional"&gt;
      &lt;pose frame=""&gt;2.32048 -2.18175 1 0 -0 0&lt;/pose&gt;
      &lt;diffuse&gt;0.5 0.5 0.5 1&lt;/diffuse&gt;
      &lt;specular&gt;0.1 0.1 0.1 1&lt;/specular&gt;
      &lt;direction&gt;0.1 0.1 -0.9&lt;/direction&gt;
      &lt;attenuation&gt;
        &lt;range&gt;20&lt;/range&gt;
        &lt;constant&gt;0.5&lt;/constant&gt;
        &lt;linear&gt;0.01&lt;/linear&gt;
        &lt;quadratic&gt;0.001&lt;/quadratic&gt;
      &lt;/attenuation&gt;
      &lt;cast_shadows&gt;1&lt;/cast_shadows&gt;
    &lt;/light&gt;
&lt;/sdf&gt;'
robot_namespace: ''
initial_pose:
  position: {x: 0.0, y: 0.0, z: 0.0}
  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
</pre>

Sorry for the bad readability of this example, let me know if it works!

**Review by davetcoleman — APPROVED** [review]

**Comment by davetcoleman** [comment]
The rosservice call is not working for me... first I had to change ``entity`` back to ``model`` but still I get errors using the updated command:

```
rosservice call /gazebo/spawn_sdf_model "entity_name: 'dirlight1'
model_xml: '<sdf version="1.5">
<light type="directional">
<pose frame="">2.32048 -2.18175 1 0 -0 0</pose>
<diffuse>0.5 0.5 0.5 1</diffuse>
<specular>0.1 0.1 0.1 1</specular>
<direction>0.1 0.1 -0.9</direction>
<attenuation>
<range>20</range>
<constant>0.5</constant>
<linear>0.01</linear>
<quadratic>0.001</quadratic>
</attenuation>
<cast_shadows>1</cast_shadows>
</light>
</sdf>'
robot_namespace: ''
initial_pose: 
position: {x: 0.0, y: 0.0, z: 0.0}
orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
```

Error:
```
ERROR: Incompatible arguments to call service:
No field name [orientation]

Service arguments are: [model_name model_xml robot_namespace initial_pose.position.x initial_pose.position.y initial_pose.position.z initial_pose.orientation.x initial_pose.orientation.y initial_pose.orientation.z initial_pose.orientation.w reference_frame]
```

I'm not sure how to nestle the ``initial_pose`` group.

**Comment by aambrosano** [comment]
That's right, I forgot to change the field name (I was trying with my local version yesterday when I wrote you, my bad). I think adding some spaces before pose and orientation should be fine. I'll try to update the comment accordingly so you can copy directly from there.

**Comment by aambrosano** [comment]
@davetcoleman did you manage to test the changes?

**Comment by davetcoleman** [comment]
@aambrosano which changes? did you update a comment? could you just repost the correct version - comment updates don't trigger emails fyi

**Comment by aambrosano** [comment]
@davetcoleman sure, sorry for the inconvenient.
The code to test the service is this:
<pre>
rosservice call /gazebo/spawn_sdf_model "model_name: 'dirlight1'
model_xml: '&lt;sdf version="1.5"&gt;
    &lt;light type="directional"&gt;
      &lt;pose frame=""&gt;2.32048 -2.18175 1 0 -0 0&lt;/pose&gt;
      &lt;diffuse&gt;0.5 0.5 0.5 1&lt;/diffuse&gt;
      &lt;specular&gt;0.1 0.1 0.1 1&lt;/specular&gt;
      &lt;direction&gt;0.1 0.1 -0.9&lt;/direction&gt;
      &lt;attenuation&gt;
        &lt;range&gt;20&lt;/range&gt;
        &lt;constant&gt;0.5&lt;/constant&gt;
        &lt;linear&gt;0.01&lt;/linear&gt;
        &lt;quadratic&gt;0.001&lt;/quadratic&gt;
      &lt;/attenuation&gt;
      &lt;cast_shadows&gt;1&lt;/cast_shadows&gt;
    &lt;/light&gt;
&lt;/sdf&gt;'
robot_namespace: ''
initial_pose:
  position: {x: 0.0, y: 0.0, z: 0.0}
  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
</pre>

**Review by davetcoleman — APPROVED** [review]

**Comment by davetcoleman** [comment]
Just tested and it works for me, thanks!

Please add a [documentation section here](http://gazebosim.org/tutorials?tut=ros_roslaunch) under spawn model that has your example for spawning lights

**Pull request merged**
